allennlp/training/trainer2.py

```diff
@@ -0,0 +1,673 @@
+
+        # Set up callback handler
+        callbacks: List[Callback] = [
+                LogTensorboard(log_batch_size_period),
+                LogTensorboardHistograms(),
+                LrsCallback(),
+                CheckpointCallback(),
+                MovingAverageCallback(),
+                Validate()
+        ]
+
+        self.handler = CallbackHandler(callbacks, self)
+
+
+    def rescale_gradients(self) -> Optional[float]:
+        return training_util.rescale_gradients(self.model, self.grad_norm)
+
+    def batch_loss(self, batch_group: List[TensorDict], for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batches and returns the ``loss`` value in the result.
+        If ``for_training`` is `True` also applies regularization penalty.
+        """
+        if self._multiple_gpu:
+            output_dict = training_util.data_parallel(batch_group, self.model, self._cuda_devices)
+        else:
+            assert len(batch_group) == 1
+            batch = batch_group[0]
+            batch = nn_util.move_to_device(batch, self._cuda_devices[0])
+            output_dict = self.model(**batch)
+
+        try:
+            loss = output_dict["loss"]
+            if for_training:
+                loss += self.model.get_regularization_penalty()
+        except KeyError:
+            if for_training:
+                raise RuntimeError("The model you are trying to optimize does not contain a"
+                                   " 'loss' key in the output of model.forward(inputs).")
+            loss = None
+
+        return loss
+
+    def _train_epoch(self, epoch: int) -> None:
+        """
+        Trains one epoch and returns metrics.
+        """
+        logger.info("Epoch %d/%d", epoch, self.num_epochs - 1)
+        peak_cpu_usage = peak_memory_mb()
+        logger.info(f"Peak CPU memory usage MB: {peak_cpu_usage}")
+        gpu_usage = []
+        for gpu, memory in gpu_memory_mb().items():
+            gpu_usage.append((gpu, memory))
+            logger.info(f"GPU {gpu} memory usage MB: {memory}")
+
+        train_loss = 0.0
+        # Set the model to "train" mode.
+        self.model.train()
+
+        num_gpus = len(self._cuda_devices)
+
+        # Get tqdm for the training batches
+        raw_train_generator = self.iterator(self.train_data,
+                                            num_epochs=1,
+                                            shuffle=self.shuffle)
+        train_generator = lazy_groups_of(raw_train_generator, num_gpus)
+        num_training_batches = math.ceil(self.iterator.get_num_batches(self.train_data)/num_gpus)
+        self.last_log = time.time()
+        last_save_time = time.time()
+
+        logger.info("Training")
+        train_generator_tqdm = Tqdm.tqdm(train_generator,
+                                         total=num_training_batches)
+        self.handler.state.cumulative_batch_size = 0
+        for batch_group in train_generator_tqdm:
```

Comment:
I am really torn on this, whether the goal should be

(1) any training regime you want should be achievable using callbacks, or
(2) it should be easy to write your own trainer, because most of the add-on functionality lives in callbacks

should you be able to train a GAN (or do multitask training, or RL) just by changing the callbacks in the trainer? or should you have to write a GAN trainer, but then you can use callbacks to get checkpointing, early stopping, etc without having to implement it all yourself.

as you can see, the latter is already pretty complicated, I think the former would be even more complicated.

for instance, the `MovingAverage` functionality made things really complex, because at several points it needs to be done before an operation (e.g. saving weights) and then undone right afterward.